File: quarkus-imperative/src/main/java/dev/nonava/techstacks/quarkus/imperative/core/domain/model/Identity.java
/*
 * Copyright 2021-2022 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package dev.nonava.techstacks.quarkus.imperative.core.domain.model;

import lombok.Value;

import java.util.UUID;

@Value
public class Identity<T extends Entity<T>> {
    UUID id;
    Class<T> type;

    public static <T extends Entity<T>> Identity<T> of(Class<T> type) {
        return new Identity<>(UUID.randomUUID(), type);
    }

    public static <T extends Entity<T>> Identity<T> of(UUID id, Class<T> type) {
        return new Identity<>(id, type);
    }

    public String asString() {
        return id.toString();
    }
}

File: ktor/src/main/kotlin/dev/nonava/techstacks/ktor/core/domain/model/Identity.kt
/*
 * Copyright 2021-2022 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package dev.nonava.techstacks.ktor.core.domain.model

import java.util.UUID
import kotlin.reflect.KClass

data class Identity<T : Entity<T>>(
    val id: UUID = UUID.randomUUID(),
    val type: KClass<T>
) {
    fun asString(): String = id.toString()
}
